### PR TITLE
Dynamic compiling in Unity Player (2019.1.6f1)

### PR DESCRIPTION
**Pull request by JanWosnitza**
Dynamically compiling an F# script in Unity3D Editor works using the following code.

    let checker = FSharpChecker.Create()
    let (_, _, asm) = checker.CompileToDynamicAssembly( [|
        "fsc.exe"
        "--target:exe"
        "--noframework"
        "--reference:" + System_DllPath
        "--reference:" + SystemIO_DllPath
        "--reference:" + FSharp_DllPath
        fsxPath
    |] )
    let main = asm |> findMainFunction // dynamic assembly doesn't set the EntryPoint
    main.Invoke(null, null)

But when building (a Windows) Player it crashes with a weird `NullReferenceException` from function `FSharp.Compiler.PrettyNaming.TryDemangleGenericNameAndPos`. The parameter `n:string` seems to be uninitialized memory (every run a random string length, most of the time negative).

This happens when [`ResolveLongIndentAsModuleOrNamespaceThen`](https://github.com/dotnet/fsharp/blob/master/src/fsharp/NameResolution.fs#L1897) is called from [here](https://github.com/dotnet/fsharp/blob/master/src/fsharp/NameResolution.fs#L2526).

Turns out the actual problem has something to do with the parameter `rest` of function `ResolveLongIndentAsModuleOrNamespaceThen` being generic. Before variable `id2` is passed into `f` it is valid, but inside [`ResolveExprLongIdentInModuleOrNamespace`](https://github.com/dotnet/fsharp/blob/master/src/fsharp/NameResolution.fs#L2295) parameter `id` is uninitialized.

The fix is to not make parameter `rest` generic, so a simple type-annotation is sufficient.

I'm pretty sure this is actually a Unity Runtime problem, but it's A LOT easier to fix it here.

PS: I know `FsiEvaluationSession` exists, but that throws an exception in it's constructor. I searched for the error and the only results I got were really old and related to some very outdated Mono version.

### CODE REVIEW

**Comment by cartermp** [comment]
Also ignore the Windows CI failures, there's an unresolved issue with our CI that we're working with others internally to address. Those failures are unrelated to this PR.

**Comment by JanWosnitza** [comment]
> I think the better change here is to just annotate rest with the type rather than do it in the match expression for the id2 value.
Done.

> Is there a Unity3D bug filed that we can cross-link as well?
Not yet, no. Don't have time for that right now. Would have to prepare a stripped down version where the bug happens.

**Comment by dsyme** [comment]
This is a really odd bug and definitely indicates a more serious bug in the UNity runtime or its version of Mono.  

Please add a comment in the code explaining why the type annotation is there, thanks.

File: src/fsharp/NameResolution.fs [hunk 1]
```diff
@@ -1894,7 +1894,7 @@ let rec ResolveLongIndentAsModuleOrNamespace sink atMostOne amap m first fullyQu
         | _ -> raze (namespaceNotFound.Force())
 
 
-let ResolveLongIndentAsModuleOrNamespaceThen sink atMostOne amap m fullyQualified (nenv: NameResolutionEnv) ad id rest isOpenDecl f =
+let ResolveLongIndentAsModuleOrNamespaceThen sink atMostOne amap m fullyQualified (nenv: NameResolutionEnv) ad id (rest: Ident list) isOpenDecl f =
```
Comment (cartermp) [inline src/fsharp/NameResolution.fs]:
```suggestion
// Note - 'rest' is annotated due to a bug currently in Unity (see: https://github.com/dotnet/fsharp/pull/7427)
let ResolveLongIndentAsModuleOrNamespaceThen sink atMostOne amap m fullyQualified (nenv: NameResolutionEnv) ad id (rest: Ident list) isOpenDecl f =
```

**Comment by JanWosnitza** [comment]
@cartermp thx for preparing :)

**Comment by cartermp** [comment]
Thanks @JanWosnitza

**Pull request merged**